allennlp/training/cross_validation.py

```diff
@@ -0,0 +1,317 @@
+        shuffle: bool = False,
+        random_state: Optional[Union[int, RandomState]] = None,
+    ) -> None:
+        super().__init__(n_splits=n_splits, shuffle=shuffle, random_state=random_state)
+
+
+@CrossValidator.register("stratified_shuffle_split")
+class StratifiedShuffleSplit(CrossValidator, sklearn.model_selection.StratifiedShuffleSplit):
+    def __init__(
+        self,
+        n_splits: int = 10,
+        test_size: Optional[Union[float, int]] = None,
+        train_size: Optional[Union[float, int]] = None,
+        random_state: Optional[Union[int, RandomState]] = None,
+    ) -> None:
+        super().__init__(
+            n_splits=n_splits, test_size=test_size, train_size=train_size, random_state=random_state
+        )
+
+
+@CrossValidator.register("time_series_split")
+class TimeSeriesSplit(CrossValidator, sklearn.model_selection.TimeSeriesSplit):
+    def __init__(self, n_splits: int = 5, max_train_size: Optional[int] = None) -> None:
+        super().__init__(n_splits=n_splits, max_train_size=max_train_size)
+
+
+logger = logging.getLogger(__name__)
+
+
+@TrainModel.register("cross_validation", constructor="from_partial_objects")
+class CrossValidateModel(TrainModel):
+    def __init__(
+        self,
+        serialization_dir: str,
+        dataset: Dataset,
+        data_loader_builder: Lazy[DataLoader],
+        model: Model,
+        trainer_builder: Lazy[Trainer],
+        cross_validator: CrossValidator,
+        batch_weight_key: str = "",
+        retrain: bool = False,
+    ) -> None:
+        data_loader = data_loader_builder.construct(dataset=dataset)
+
+        # The "main" trainer. It's used to pass to super and when `retrain` option is set.
+        #
+        # We don't need to pass `serialization_dir` and `local_rank` here, because they will have
+        # been passed through the trainer by from_params already, because they were keyword
+        # arguments to construct this class in the first place.
+        trainer = trainer_builder.construct(model=model, data_loader=data_loader)
+
+        super().__init__(serialization_dir=serialization_dir, model=model, trainer=trainer)
+
+        self.dataset = dataset
+        self.data_loader_builder = data_loader_builder
+        self.trainer_builder = trainer_builder
+        self.batch_weight_key = batch_weight_key
+        self.cross_validator = cross_validator
+        self.retrain = retrain
+
+    @overrides
+    def run(self) -> Dict[str, Any]:
+        metrics_by_fold = []
+
+        n_splits = self.cross_validator.get_n_splits(self.dataset)
+
+        for fold_index, (train_indices, test_indices) in enumerate(
+            self.cross_validator(self.dataset)
+        ):
+            logger.info(f"Fold {fold_index}/{n_splits - 1}")
+
+            serialization_dir = os.path.join(self.serialization_dir, f"fold_{fold_index}")
+            os.makedirs(serialization_dir, exist_ok=True)
+
+            train_dataset = Subset(self.dataset, train_indices)
```

Comment:
This fails if there's a bucket sampler, because it needs the `Vocabulary` from the `Dataset`, but `Subset` doesn't have one. Thoughts? I see 2 options: the sampler having a `Vocabulary` arg instead of a `Dataset` (it only uses it for that), or implementing our own `Subset`.

Reply:
There's also just doing: `train_dataset.vocab = self.dataset.vocab`

Reply:
I don't understand this enough to answer this question.  I'd need more detail / to read more, and I don't have time for that right now.

Reply:
`BucketBatchSampler` needs a special `Dataset`: one that has a `vocab` attribute. `Subset` doesn't.

https://github.com/allenai/allennlp/blob/644ef22e90c8493af8e14e81a1e10839a8f0b62a/allennlp/data/samplers/bucket_batch_sampler.py#L69

Reply:
The problem is that torch's `Dataset`s don't have a `vocab` field, and they don't have a `index_with` method either. So whenever you derive a new `Dataset` from an existing `AllennlpDataset`, using torch methods, those two things get lost.

I think what we should do is just use straight torch `Dataset` everywhere, move the vocab around manually (or maybe inside the instances?), and have a function `index_dataset_with_vocab(dataset, vocab)` which we call instead of `Dataset.index_with`.

I also think that should not be part of this PR. For this PR, `train_dataset.vocab = self.dataset.vocab` is fine. Let's open another issue for the follow-up work item. Many small PRs are better than one huge one.

Reply:
I also definitely agree on not doing this in this PR.  I think what you're suggesting, @dirkgr, is having the sampler / dataloader own the vocab, instead of the dataset?  That would probably work, but I'd have to dig through the API to be sure.

Reply:
Agree on not doing it in this PR. Wanted to open up the discussion. What if `BucketBatchSampler` receives a `Vocabulary` as an arg? That'd solve it.

Reply:
Also, note `BucketBatchSampler` doesn't just accept any `Dataset`, as internally it assumes it gives `Instance`s.